helix-term/src/commands/lsp.rs

```diff
@@ -424,81 +497,117 @@ impl ui::menu::Item for lsp::CodeActionOrCommand {
 pub fn code_action(cx: &mut Context) {
     let (view, doc) = current!(cx.editor);
 
-    let language_server = language_server!(cx.editor, doc);
-
     let selection_range = doc.selection(view.id).primary();
-    let offset_encoding = language_server.offset_encoding();
 
-    let range = range_to_lsp_range(doc.text(), selection_range, offset_encoding);
+    // this ensures, that a previously opened menu that doesn't have anything to do with this command will be replaced with a new menu
+    let code_actions_menu_open = Arc::new(Mutex::new(false));
 
-    let future = language_server.code_actions(
-        doc.identifier(),
-        range,
-        // Filter and convert overlapping diagnostics
-        lsp::CodeActionContext {
-            diagnostics: doc
-                .diagnostics()
-                .iter()
-                .filter(|&diag| {
-                    selection_range
-                        .overlaps(&helix_core::Range::new(diag.range.start, diag.range.end))
-                })
-                .map(|diag| diagnostic_to_lsp_diagnostic(doc.text(), diag, offset_encoding))
-                .collect(),
-            only: None,
-        },
-    );
+    let mut requests = Vec::new();
 
-    cx.callback(
-        future,
-        move |editor, compositor, response: Option<lsp::CodeActionResponse>| {
-            let actions = match response {
-                Some(a) => a,
-                None => return,
-            };
-            if actions.is_empty() {
-                editor.set_status("No code actions available");
-                return;
-            }
+    for language_server in doc.language_servers_with_feature(LanguageServerFeature::CodeAction) {
+        let offset_encoding = language_server.offset_encoding();
+        let range = range_to_lsp_range(doc.text(), selection_range, offset_encoding);
+
+        requests.push((
+            language_server.code_actions(
+                doc.identifier(),
+                range,
+                // Filter and convert overlapping diagnostics
+                lsp::CodeActionContext {
+                    diagnostics: doc
+                        .diagnostics()
+                        .iter()
+                        .filter(|&diag| {
+                            selection_range
+                                .overlaps(&helix_core::Range::new(diag.range.start, diag.range.end))
+                        })
+                        .map(|diag| diagnostic_to_lsp_diagnostic(doc.text(), diag, offset_encoding))
+                        .collect(),
+                    only: None,
+                },
+            ),
+            offset_encoding,
+            language_server.id(),
+        ));
+    }
+
+    for (future, offset_encoding, lsp_id) in requests {
```

Comment:
I'm noticing this pattern in several places in this PR where you collect everything into a `Vec` and then immediately iterate over it. Why are we doing this? Why not just do it all in the same loop?

Reply:
This has to do with borrow-checking, in fact if I remember right (well it's a few months since I wrote this...), I also tried doing a lot more with iterators etc. but borrow-checking and potential workarounds convoluted the code way more than just allocating a temporary `Vec` for this. I think an allocation for keeping the code more readable is a better tradeoff IMHO. Sometimes it's not avoidable with an Iterator anyway (i.e. `Box<dyn Iterator>`). I think since this (and most other code where I did this) is executed rarely (in this case just when a `code_action` command is requested), I leaned for more readable instead of potentially better (technically) or faster code.

Reply:
Makes sense, thanks for the explanation!